src/CalldataList.sol

```diff
@@ -0,0 +1,218 @@
+    /// @param startIndex the start index of the calldata
+    /// @param endIndex the end index of the calldata
+    /// @param data the calldata that is stored
+    event CalldataAdded(
+        address indexed contractAddress,
+        bytes4 indexed selector,
+        uint16 startIndex,
+        uint16 endIndex,
+        bytes data
+    );
+
+    /// @notice event emitted when a calldata check is removed
+    /// @param contractAddress the address of the contract that the calldata check is removed from
+    /// @param selector the function selector of the function that the calldata check is removed from
+    /// @param startIndex the start index of the calldata
+    /// @param endIndex the end index of the calldata
+    /// @param data the calldata that is removed
+    event CalldataRemoved(
+        address indexed contractAddress,
+        bytes4 indexed selector,
+        uint16 startIndex,
+        uint16 endIndex,
+        bytes data
+    );
+
+    /// @notice struct used to store the start and end index of the calldata
+    /// and the calldata itself.
+    /// Once the calldata is stored, it can be used to check if the calldata
+    /// conforms to the expected values.
+    struct Index {
+        uint16 startIndex;
+        uint16 endIndex;
+        bytes data;
+    }
+
+    /// @notice struct used to store the calldata checks and the maximum
+    /// amount of native tokens to be sent in a single call to a contract
+    struct CallRestriction {
+        Index[] calldataChecks;
+        uint256 maxValue;
+    }
+
+    /// @notice mapping of contract address to function selector to Index struct
+    mapping(
+        address contractAddress
+            => mapping(bytes4 selector => CallRestriction calldataChecks)
+    ) public calldataList;
+
+    /// @notice get the calldata checks for a specific contract and function selector
+    function getCalldataChecks(address contractAddress, bytes4 selector)
+        public
+        view
+        returns (Index[] memory)
+    {
+        return calldataList[contractAddress][selector].calldataChecks;
+    }
+
+    /// @notice check if the calldata conforms to the expected values
+    /// extracts indexes and checks that the data within the indexes
+    /// matches the expected data
+    /// @param contractAddress the address of the contract that the calldata check is applied to
+    /// @param value the amount of native asset sent with the call
+    /// @param data the calldata to check
+    function checkCalldata(
+        address contractAddress,
+        uint256 value,
+        bytes memory data
+    ) public view {
+        bytes4 selector = data.getFunctionSignature();
+
+        Index[] storage calldataChecks =
+            calldataList[contractAddress][selector].calldataChecks;
+
+        require(calldataChecks.length > 0, "No calldata checks found");
+        require(
+            value <= calldataList[contractAddress][selector].maxValue,
+            "Value exceeds maximum"
```

Comment:
In CalldataList we can update the revert messages similar to Timelock and TimeRestricted. `CalldataList: Value exceeds maximum`